package/MDAnalysis/lib/c_distances.pyx

```diff
@@ -315,3 +317,307 @@ def contact_matrix_pbc(coord, sparse_contacts, box, cutoff):
+
+            elif hi_y:
+                for j in range(3):
+                    output[p, j] = coord[j] - shiftX[j] - shiftY[j]
+                indices[p] = i
+                p += 1
+
+                if lo_z:
+                    for j in range(3):
+                        output[p, j] = coord[j] - shiftX[j] - shiftY[j] + shiftZ[j]
+                    indices[p] = i
+                    p += 1
+
+                elif hi_z:
+                    for j in range(3):
+                        output[p, j] = coord[j] - shiftX[j] - shiftY[j] - shiftZ[j]
+                    indices[p] = i
+                    p += 1
+
+            if lo_z:
+                for j in range(3):
+                    output[p, j] = coord[j] - shiftX[j] + shiftZ[j]
+                indices[p] = i
+                p += 1
+
+            elif hi_z:
+                for j in range(3):
+                    output[p, j] = coord[j] - shiftX[j] - shiftZ[j]
+                indices[p] = i
+                p += 1
+
+        if lo_y:
+            for j in range(3):
+                output[p, j] = coord[j] + shiftY[j]
+            indices[p] = i
+            p += 1
+
+            if lo_z:
+                for j in range(3):
+                    output[p, j] = coord[j] + shiftY[j] + shiftZ[j]
+                indices[p] = i
+                p += 1
+
+            elif hi_z:
+                for j in range(3):
+                    output[p, j] = coord[j] + shiftY[j] - shiftZ[j]
+                indices[p] = i
+                p += 1
+
+        elif hi_y:
+            for j in range(3):
+                output[p, j] = coord[j] - shiftY[j]
+            indices[p] = i
+            p += 1
+
+            if lo_z:
+                for j in range(3):
+                    output[p, j] = coord[j] - shiftY[j] + shiftZ[j]
+                indices[p] = i
+                p += 1
+
+            elif hi_z:
+                for j in range(3):
+                    output[p, j] = coord[j] - shiftY[j] - shiftZ[j]
+                indices[p] = i
+                p += 1
+        
+        if lo_z:
+            for j in range(3):
+                output[p, j] = coord[j] + shiftZ[j]
+            indices[p] = i
+            p += 1
+
+        elif hi_z:
+            for j in range(3):
+                output[p, j] = coord[j] - shiftZ[j]
+            indices[p] = i
+            p += 1
+    return numpy.array(output[:p]), numpy.array(indices[:p])
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef float _dot(float * a, float * b):
+    """Return dot product of two sequences in range."""
+    cdef ssize_t n
+    cdef float sum1
+
+    sum1 = 0.0
+    for n in range(3):
+        sum1 += a[n] * b[n]
+    return sum1
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef float[:] _cross(float * a, float * b):
+    """
+    Calculates the cross product between vectors
+    given by pointers a and b
+    """
+    cdef ssize_t n
+    cdef float[:] result = numpy.zeros((3,), dtype=numpy.float32)
```

Comment:
in general you don't want any numpy calls inside a cdef function, this should all just be pure C.  We only want numpy calls at the start of a function (to prepare things) and at the end (to correctly format things for the user)